File: 剑指offer/54 二叉搜索树的第k大节点.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int kthLargest(TreeNode root, int k) {
        ArrayList<Integer> arr = new ArrayList<>();
        add(root, arr);
        return arr.get(arr.size() - k);
    }
    
    private void add(TreeNode root, ArrayList<Integer> arr) {
        if (null == root) return;
        add(root.left, arr);
        arr.add(root.val);
        add(root.right, arr);
    }
}

class Solution {
    private int idx = 0;
    private int res = 0;
    
    public int kthLargest(TreeNode root, int k) {
        int sum = nodeSum(root);
        find(root, sum - k + 1);
        return res;
    }
    
    private int nodeSum(TreeNode root) {
        if (null == root) return 0;
        int sum = 1;
        sum += nodeSum(root.left);
        sum += nodeSum(root.right);
        return sum;
    }
    
    private void find(TreeNode root, int aim) {
        if (null == root) return;
        find(root.left, aim);
        ++idx;
        if (idx == aim) res = root.val;
        find(root.right, aim);
    }
}

File: 剑指offer/54 二叉搜索树的第k大节点.cpp
class Solution {
public:
    int kthLargest(TreeNode* root, int k) {
        int remainCount = countNode(root) - k + 1;
        int res = 0;
        findRes(root, remainCount, res);
        return res;
    }

    int countNode(TreeNode* root) {
        if (root == nullptr) return 0;
        return countNode(root->left) + countNode(root->right) + 1;
    }

    void findRes(TreeNode* root, int& remainCount, int& res) {
        if (root == nullptr) return;
        findRes(root->left, remainCount, res);
        if (--remainCount == 0) {
            res = root->val;
            return;
        }
        findRes(root->right, remainCount, res);
    }
};